### Movie Recommendation Model 
The objective of this project is to predict users' ratings on movies and recommend each user their top 10 movies with highest rating. We will use dataset from [MovieLens](https://grouplens.org/datasets/movielens/latest/) to train the model.

Collaborative filtering(CF) is commonly used for recommender systems. In this notebook, Alternating Least Squares (ALS) algorithm in spark.ml packages was applied to develope a recommendation model based on Spark APIs. ALS allows us to describe users and products by a small set of latent factors, which can be used to predict ratings. Compared to item based or user based CF recommender, instead of recommendering similar product, ALS can recommenders product personally fitted for a given user.

## Data ETL and Data Exploration Analysis

In [3]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("moive analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [4]:
movies = spark.read.load("/FileStore/tables/movies.csv", format='csv', header = True)
ratings = spark.read.load("/FileStore/tables/ratings.csv", format='csv', header = True)
movies.createOrReplaceTempView("sqlmovies")
ratings.createOrReplaceTempView("sqlratings")

In [5]:
display(movies)

movieId title genres 1 Toy Story (1995) Adventure|Animation|Children|Comedy|Fantasy 2 Jumanji (1995) Adventure|Children|Fantasy 3 Grumpier Old Men (1995) Comedy|Romance 4 Waiting to Exhale (1995) Comedy|Drama|Romance 5 Father of the Bride Part II (1995) Comedy 6 Heat (1995) Action|Crime|Thriller 7 Sabrina (1995) Comedy|Romance 8 Tom and Huck (1995) Adventure|Children 9 Sudden Death (1995) Action 10 GoldenEye (1995) Action|Adventure|Thriller 11 American President, The (1995) Comedy|Drama|Romance 12 Dracula: Dead and Loving It (1995) Comedy|Horror 13 Balto (1995) Adventure|Animation|Children 14 Nixon (1995) Drama 15 Cutthroat Island (1995) Action|Adventure|Romance 16 Casino (1995) Crime|Drama 17 Sense and Sensibility (1995) Drama|Romance 18 Four Rooms (1995) Comedy 19 Ace Ventura: When Nature Calls (1995) Comedy 20 Money Train (1995) Action|Comedy|Crime|Drama|Thriller 21 Get Shorty (1995) Comedy|Crime|Thriller 22 Copycat (1995) Crime|Drama|Horror|Mystery|Thriller 23 Assassins (1995) Action|Crime|Thriller 24 Powder (1995) Drama|Sci-Fi 25 Leaving Las Vegas (1995) Drama|Romance 26 Othello (1995) Drama 27 Now and Then (1995) Children|Drama 28 Persuasion (1995) Drama|Romance 29 City of Lost Children, The (Cité des enfants perdus, La) (1995) Adventure|Drama|Fantasy|Mystery|Sci-Fi 30 Shanghai Triad (Yao a yao yao dao waipo qiao) (1995) Crime|Drama 31 Dangerous Minds (1995) Drama 32 Twelve Monkeys (a.k.a. 12 Monkeys) (1995) Mystery|Sci-Fi|Thriller 33 Wings of Courage (1995) Adventure|Romance|IMAX 34 Babe (1995) Children|Drama 35 Carrington (1995) Drama|Romance 36 Dead Man Walking (1995) Crime|Drama 37 Across the Sea of Time (1995) Documentary|IMAX 38 It Takes Two (1995) Children|Comedy 39 Clueless (1995) Comedy|Romance 40 Cry, the Beloved Country (1995) Drama 41 Richard III (1995) Drama|War 42 Dead Presidents (1995) Action|Crime|Drama 43 Restoration (1995) Drama 44 Mortal Kombat (1995) Action|Adventure|Fantasy 45 To Die For (1995) Comedy|Drama|Thriller 46 How to Make an American Quilt (1995) Drama|Romance 47 Seven (a.k.a. Se7en) (1995) Mystery|Thriller 48 Pocahontas (1995) Animation|Children|Drama|Musical|Romance 49 When Night Is Falling (1995) Drama|Romance 50 Usual Suspects, The (1995) Crime|Mystery|Thriller 51 Guardian Angel (1994) Action|Drama|Thriller 52 Mighty Aphrodite (1995) Comedy|Drama|Romance 53 Lamerica (1994) Adventure|Drama 54 Big Green, The (1995) Children|Comedy 55 Georgia (1995) Drama 56 Kids of the Round Table (1995) Adventure|Children|Comedy|Fantasy 57 Home for the Holidays (1995) Drama 58 Postman, The (Postino, Il) (1994) Comedy|Drama|Romance 59 Confessional, The (Confessionnal, Le) (1995) Drama|Mystery 60 Indian in the Cupboard, The (1995) Adventure|Children|Fantasy 61 Eye for an Eye (1996) Drama|Thriller 62 Mr. Holland's Opus (1995) Drama 63 Don't Be a Menace to South Central While Drinking Your Juice in the Hood (1996) Comedy|Crime 64 Two if by Sea (1996) Comedy|Romance 65 Bio-Dome (1996) Comedy 66 Lawnmower Man 2: Beyond Cyberspace (1996) Action|Sci-Fi|Thriller 67 Two Bits (1995) Drama 68 French Twist (Gazon maudit) (1995) Comedy|Romance 69 Friday (1995) Comedy 70 From Dusk Till Dawn (1996) Action|Comedy|Horror|Thriller 71 Fair Game (1995) Action 72 Kicking and Screaming (1995) Comedy|Drama 73 Misérables, Les (1995) Drama|War 74 Bed of Roses (1996) Drama|Romance 75 Big Bully (1996) Comedy|Drama 76 Screamers (1995) Action|Sci-Fi|Thriller 77 Nico Icon (1995) Documentary 78 Crossing Guard, The (1995) Action|Crime|Drama|Thriller 79 Juror, The (1996) Drama|Thriller 80 White Balloon, The (Badkonake sefid) (1995) Children|Drama 81 Things to Do in Denver When You're Dead (1995) Crime|Drama|Romance 82 Antonia's Line (Antonia) (1995) Comedy|Drama 83 Once Upon a Time... When We Were Colored (1995) Drama|Romance 84 Last Summer in the Hamptons (1995) Comedy|Drama 85 Angels and Insects (1995) Drama|Romance 86 White Squall (1996) Action|Adventure|Drama 87 Dunston Checks In (1996) Children|Comedy 88 Black Sheep (1996) Comedy 89 

In [6]:
display(ratings)

userId movieId rating timestamp 1 307 3.5 1256677221 1 481 3.5 1256677456 1 1091 1.5 1256677471 1 1257 4.5 1256677460 1 1449 4.5 1256677264 1 1590 2.5 1256677236 1 1591 1.5 1256677475 1 2134 4.5 1256677464 1 2478 4.0 1256677239 1 2840 3.0 1256677500 1 2986 2.5 1256677496 1 3020 4.0 1256677260 1 3424 4.5 1256677444 1 3698 3.5 1256677243 1 3826 2.0 1256677210 1 3893 3.5 1256677486 2 170 3.5 1192913581 2 849 3.5 1192913537 2 1186 3.5 1192913611 2 1235 3.0 1192913585 2 1244 3.0 1192913551 2 1296 4.5 1192913608 2 1663 3.0 1192913565 2 1962 4.0 1192913533 2 2108 3.5 1192913541 2 2243 4.5 1192913604 2 2352 4.0 1192913575 2 2707 3.5 1192913600 2 2746 4.0 1192913548 2 2915 3.5 1192913562 2 3363 4.0 1192913596 3 640 3.0 945141610 3 828 4.0 945141610 3 960 3.0 945141407 3 1221 4.0 945141407 3 1321 4.0 945141407 3 1645 4.0 945141611 3 1825 3.0 945141611 3 1985 2.0 945141611 3 2024 3.0 945141611 3 2028 5.0 945141611 3 3171 4.0 945141407 4 1 4.0 1113765937 4 2 4.0 1113767306 4 5 2.0 1123990453 4 6 4.5 1113767242 4 10 4.0 1113765995 4 11 3.5 1113767280 4 16 4.0 1113765691 4 19 2.0 1113767324 4 20 2.5 1145937918 4 23 3.0 1115782385 4 25 4.5 1113767215 4 31 3.0 1113796669 4 32 4.5 1113767082 4 34 3.5 1113767134 4 36 3.5 1113767240 4 39 2.5 1113767219 4 41 4.5 1123990211 4 44 1.0 1123989796 4 45 2.0 1117080123 4 47 5.0 1113767129 4 50 4.5 1113767113 4 61 3.0 1114050467 4 62 3.5 1135543206 4 65 0.5 1115782387 4 66 0.5 1145937915 4 70 3.0 1117080064 4 76 2.0 1113766115 4 85 4.0 1115782362 4 95 4.5 1113767211 4 100 4.0 1115782366 4 104 3.0 1113765671 4 110 4.5 1113767070 4 141 4.5 1113767207 4 145 4.0 1115782372 4 150 4.0 1113765768 4 151 4.5 1123990462 4 153 2.0 1113767107 4 158 0.5 1127946524 4 160 2.5 1113767377 4 161 4.0 1113765802 4 163 2.5 1113765975 4 165 4.5 1113767110 4 169 0.5 1113796675 4 170 3.0 1114050521 4 172 3.5 1113796672 4 173 3.5 1114050212 4 175 3.5 1127946774 4 180 4.5 1113797176 4 181 0.5 1127946789 4 185 3.0 1113767187 4 186 2.0 1113766782 4 193 0.5 1113766123 4 198 4.5 1135543037 4 204 3.5 1145937771 4 216 2.0 1115782393 4 223 4.5 1113767287 4 231 0.5 1113767153 4 253 4.5 1113767185 4 255 0.5 1113796828 4 256 2.5 1127946576 4 260 4.5 1127947466 4 267 3.5 1114050525 4 288 4.5 1113767252 4 292 3.5 1113767157 4 293 4.5 1113795816 4 296 5.0 1113767056 4 315 2.0 1113796665 4 316 5.0 1113767120 4 317 2.0 1113767337 4 318 5.0 1113766505 4 327 1.5 1113766111 4 329 3.5 1113766260 4 332 2.0 1114050781 4 333 1.5 1115782380 4 338 2.5 1117080407 4 339 0.5 1123990368 4 344 3.0 1113767090 4 349 5.0 1113765966 4 350 2.0 1113767328 4 353 4.0 1113767406 4 356 3.5 1113767059 4 361 3.5 1113766326 4 364 4.0 1113766585 4 367 2.5 1113765954 4 368 3.5 1113765650 4 370 2.0 1123990502 4 377 3.5 1113767088 4 379 1.5 1114050471 4 380 4.0 1113765930 4 382 3.0 1127946804 4 405 3.0 1114050485 4 410 1.0 1123990391 4 420 2.0 1123990435 4 423 3.0 1117080440 4 432 2.5 1123990449 4 434 1.0 1113767171 4 435 2.0 1115782018 4 440 4.0 1113767269 4 442 2.0 1113767333 4 454 2.0 1113767181 4 455 2.0 1117080429 4 457 4.5 1113767064 4 466 2.5 1123990512 4 471 4.5 1113796025 4 474 3.5 1127946566 4 480 4.0 1113767062 4 481 4.0 1204592835 4 485 0.5 1113796307 4 494 4.0 1113766249 4 500 3.0 1113767148 4 514 4.5 1145937691 4 517 3.5 1117080425 4 520 3.5 1123990009 4 524 4.0 1115781846 4 527 4.5 1113767035 4 539 0.5 1113767169 4 541 4.5 1127947584 4 546 0.5 1115782390 4 552 3.5 1114050135 4 586 2.5 1117080127 4 587 2.0 1113767167 4 588 2.5 1113797093 4 589 4.5 1113767073 4 590 4.0 1113766228 4 592 4.0 1113767067 4 593 4.5 1113766512 4 595 2.5 1115781822 4 597 3.0 1113767146 4 608 5.0 1113767086 4 610 4.0 1113766854 4 647 3.5 1113766255 4 648 4.0 1113767094 4 653 3.5 1113766757 4 724 3.0 1117080436 4 733 5.0 1113767139 4 736 3.5 1113767125 4 741 4.0 1113795914 4 762 3.5 1117080073 4 778 3.5 1113767358 4 780 4.0 1113766216 4 784 0.5 1114050139 4 785 4.0 1114050086 4 786 4.0 1113765607 4 788 1.0 1123990395 4 798 3.0 1115782316 4 805 4.0 1113796656 4 818 0.5 1145937873 4 832 4.

In [7]:
ratings.select('rating').summary().show()

+-------+------------------+
summary| rating|
+-------+------------------+
 count| 27753444|
 mean|3.5304452124932677|
 stddev| 1.066352750231989|
 min| 0.5|
 25%| 3.0|
 50%| 3.5|
 75%| 4.0|
 max| 5.0|
+-------+------------------+

In [8]:
display(ratings.select('rating').groupBy('rating').count().orderBy('rating'))

rating count 0.5 442388 1.0 886233 1.5 441354 2.0 1850627 2.5 1373419 3.0 5515668 3.5 3404360 4.0 7394710 4.5 2373550 5.0 4071135

In [9]:
print('Total number of users rated on movies:', ratings.select('userId').distinct().count())
print('Total number of movies rated by users:', ratings.select('movieId').distinct().count())
print('Total number of movies: ', movies.select('movieId').count())
print('Number of movies not rated by user: ', spark.sql("select * from sqlmovies where movieId not in (select distinct(movieId) from sqlratings)").count())

Total number of users rated on movies: 283228
Total number of movies rated by users: 53889
Total number of movies: 58098
Number of movies not rated by user: 4209

In [10]:
tmp1 = ratings.groupBy("userID").count().toPandas()['count'].min()
tmp2 = ratings.groupBy("movieId").count().toPandas()['count'].min()
print('For the users that rated movies and the movies that were rated:')
print('Minimum number of ratings per user is {}'.format(tmp1))
print('Minimum number of ratings per movie is {}'.format(tmp2))

For the users that rated movies and the movies that were rated:
Minimum number of ratings per user is 1
Minimum number of ratings per movie is 1

In [11]:
tmp1 = sum(ratings.groupBy("movieId").count().toPandas()['count'] == 1)
tmp2 = ratings.select('movieId').distinct().count()
print('{} out of {} movies are rated by only one user'.format(tmp1, tmp2))

10155 out of 53889 movies are rated by only one user

List of Movie Genres:

In [13]:
%sql select genres, count(genres) from (select title, explode(split(genres, '[|]')) as genres from sqlmovies) group by genres order by 2 DESC

genres count(genres) Drama 24144 Comedy 15956 Thriller 8216 Romance 7412 Action 7130 Horror 5555 Documentary 5118 Crime 5105 (no genres listed) 4266 Adventure 4067 Sci-Fi 3444 Mystery 2773 Children 2749 Animation 2663 Fantasy 2637 War 1820 Western 1378 Musical 1113 Film-Noir 364 IMAX 197

###Build Model

In [15]:
# import package
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import CrossValidator,ParamGridBuilder
from pyspark.sql import functions

In [16]:
# Data type convert for model training
from pyspark.sql.types import IntegerType, FloatType
ratings = ratings.withColumn("userId", ratings["userId"].cast(IntegerType()))
ratings = ratings.withColumn("movieId", ratings["movieId"].cast(IntegerType()))
ratings = ratings.withColumn("rating", ratings["rating"].cast(FloatType()))

In [17]:
ratings = ratings.drop('timestamp')

Split 20% of data for testing

In [19]:
(training,test) = ratings.randomSplit([0.8,0.2])

In [20]:
print('training set size: ', training.count())
print('testing set size: ', test.count())

training set size: 22204865
testing set size: 5548579

In [21]:
#cache dataset to memory for faster model training
training.cache()
test.cache()

Out[ 162 ]: DataFrame[userId: int, movieId: int, rating: float]

Since the model is trained based on a users-movies matrix and aims to fill in missing value in the matrix, it is very important for us to know the sparsity of the martix. Theoretically, denser matrix will give us better result because we feed more information to the model.

In [23]:
numOfRatings = training.select('rating').count()
numOfMovies = training.select('movieId').distinct().count()
numOfUsers = training.select('userId').distinct().count()
print('The sparsity of user-movie matrix: {:.2f} %'.format(numOfRatings / (numOfMovies * numOfUsers) * 100))

The sparsity of user-movie matrix: 0.15 %

Here we can tune model using pyspark.ml.tuning packages, however, due to limitted computional power, we will define our own hyperparameters.

In [25]:
#Tune model using ParamGridBuilder, didnot use
param_grid = ParamGridBuilder()\
    .addGrid(als.rank, [6, 8, 10, 12]) \
    .addGrid(als.maxIter, [10,15,20])\
    .addGrid(als.regParam, [0.05, 0.1, 0.2])\
    .build()
#Define evaluator as RMSE
evaluator=RegressionEvaluator(metricName='rmse',labelCol='rating',predictionCol='prediction')

# Build Cross validation 
cv = CrossValidator(estimator=als, estimatorParamMaps=param_grid, evaluator=evaluator, numFolds=3) 

#Fit ALS model to training data
model=cv.fit(training)

#Extract best model from the tuning exercise using ParamGridBuilder
best_model=model.bestModel 

###Train Model

When making predictions using an ALSModel, it is common to encounter users and/or items in the test dataset that were not present during training the model. During cross-validation, the data is split between training and evaluation sets so the model will predict the ratings of unseen users or items as N/A, which will affect the model scores. Here we set coldStartStrategy="drop".

In [27]:
als = ALS(maxIter=20, regParam=0.2, rank=6, userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")
import time
start_time = time.time()
model = als.fit(training)
print ('Total Runtime: {:.2f} seconds'.format(time.time() - start_time))

Total Runtime: 251.18 seconds

In [28]:
# Evaluate the model by computing the RMSE on the test data
predictions = model.transform(test)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse = evaluator.evaluate(predictions)
print("Root-mean-square error: {:.2f}".format(rmse))
print('Normalized error:  {:.2f} %'.format(rmse/5*100))

Root-mean-square error: 0.87
Normalized error: 17.40 %

In [29]:
display(predictions)

userId movieId rating prediction 107339 148 4.0 3.1317124 253535 148 4.0 2.991921 52620 148 1.0 2.7181563 60382 148 4.0 3.3250985 275860 148 3.0 2.615859 8350 148 4.0 2.8657644 51571 148 3.0 2.966929 52772 148 3.0 3.4331298 73492 148 0.5 2.5985732 60950 148 2.0 2.4365332 64681 148 1.5 2.561519 111686 148 4.0 2.4847891 111694 148 5.0 3.3523078 232773 148 4.0 3.053217 68789 148 2.0 2.5272589 84032 148 3.0 1.4473698 47895 148 3.0 2.7382412 255673 148 3.0 2.7577786 40716 148 2.5 2.2998238 6826 148 3.0 2.5411403 160910 148 3.0 3.0234928 107574 148 1.0 2.9889588 146260 148 3.0 2.7993848 112751 148 2.0 2.9147124 139582 148 3.0 2.838426 50585 148 3.0 2.4108748 37350 148 4.0 2.7179804 123100 148 3.0 2.6076653 277144 148 4.0 3.548147 115423 148 4.0 3.0931342 75370 148 5.0 3.0703902 280788 148 3.0 2.5360253 227590 148 5.0 3.407837 204845 148 3.0 3.1081414 232701 148 3.0 3.2149522 9051 148 1.0 2.789976 7051 148 5.0 2.7950501 48093 148 1.0 2.8045259 33705 148 3.0 2.6489375 159331 148 3.0 2.5376203 39682 148 3.0 2.5540385 62861 148 4.5 3.0763621 222373 148 5.0 3.2728221 16066 148 3.0 2.6953342 270252 148 4.0 3.0529032 35187 148 3.0 2.5470376 55508 148 2.0 2.9001484 211042 148 4.0 3.4604003 147763 148 2.0 2.2439532 58025 148 3.0 2.6646893 156955 148 3.0 2.4882126 182536 148 2.0 3.0491405 33809 148 1.0 3.004673 25365 148 3.0 2.8285656 17671 148 3.0 3.6217794 202056 148 1.0 2.911217 149820 148 1.0 2.2205958 237193 148 5.0 2.200045 175141 148 2.0 2.83078 60802 148 3.0 2.7610226 7410 148 3.0 2.5104246 152161 148 4.0 2.8485744 262181 148 4.5 2.46249 111055 148 3.0 2.8927288 167531 148 3.0 2.5326672 91374 148 3.0 2.4097965 95921 148 4.0 2.4143987 136537 148 2.0 2.964343 35578 148 3.0 2.8852897 129620 463 3.0 2.9081936 242758 463 3.0 2.690107 95586 463 3.0 2.804352 260317 463 2.5 2.0470648 89695 463 2.0 2.7331257 280540 463 2.0 2.192156 44008 463 1.0 2.4207087 19679 463 3.0 2.6887505 36898 463 2.0 2.9809756 171473 463 3.0 2.6136193 146406 463 1.5 1.6890349 88355 463 3.0 2.1440008 113435 463 3.0 2.955822 170203 463 2.5 2.560155 187543 463 3.0 2.5152912 215925 463 2.0 2.8418229 265145 463 2.0 2.7925663 161487 463 3.0 2.6930099 20571 463 2.0 2.8947136 94321 463 4.0 2.9245143 109343 463 2.0 2.377739 234019 463 3.0 2.997088 247628 463 2.0 2.4805837 139062 463 3.5 2.9884043 64180 463 3.0 2.7265625 119455 463 2.5 2.3615003 154072 463 2.0 2.605064 11752 463 4.0 2.914765 66096 463 1.0 2.1548142 214935 463 1.5 1.9333007 157870 463 3.0 2.7817037 7427 463 3.0 2.8114252 235466 463 3.0 2.658147 190797 463 3.0 3.2009664 110391 463 3.5 2.9947944 277244 463 4.0 3.202173 232489 463 3.0 2.8560271 61801 463 5.0 3.30689 122978 463 3.0 3.338471 79738 463 4.0 2.4821806 273415 463 2.0 2.8041058 55431 463 4.0 2.922832 274146 463 3.0 2.3942738 62064 463 2.5 2.743805 130756 463 2.0 2.5336823 148114 463 3.0 2.4457178 208904 463 2.0 2.88465 154580 463 3.0 2.7156124 102416 463 3.0 2.5827425 249542 463 3.0 1.9348797 122149 463 3.0 2.6925616 111958 463 2.0 2.183468 155764 463 3.0 2.7385306 210747 463 4.0 3.1225433 23519 463 3.0 2.7341597 45155 463 2.0 2.4477682 69046 463 2.0 2.7158942 181811 463 3.0 2.5180435 155809 463 3.0 2.6831026 41734 463 4.0 3.284451 77609 463 2.0 2.1427817 136238 463 3.0 2.7287292 227694 463 3.0 2.6415906 187026 463 1.5 2.3647857 171746 463 3.0 3.432447 178744 463 1.0 2.583963 205308 463 3.0 2.875381 282953 463 4.0 3.264173 41791 463 1.0 2.3853762 180927 463 1.0 2.3955345 159968 463 2.0 2.9321065 195338 463 4.0 2.796749 67530 463 5.0 3.5836744 107143 463 3.0 2.5265682 164074 463 5.0 3.1305025 48736 463 3.0 2.6887488 265521 463 2.0 2.6117134 139053 463 4.0 2.7975304 24934 463 2.0 2.0657687 78982 463 0.5 2.1503363 172047 463 2.0 1.9423523 26096 463 1.0 2.3422484 184085 463 2.5 2.1863794 219351 463 2.0 3.0018125 270436 463 1.0 2.7048059 9363 463 1.0 2.1594722 206404 463 4.0 2.8906941 242289 463 3.0 3.0971723 112685 463 5.0 3.0013857 35278 463 2.0 2.6625857 151465 463 2.0 2.4041905 267167 463 2.0 2.5430083 93820 463 1.0 2.6497707 83755 463 2.0 2.5015774 10362 

### Apply model to recommend movies

In [31]:
# Generate top 10 movie recommendations for each user
userRecs = model.recommendForAllUsers(10)
users = ratings.select(als.getUserCol()).distinct().limit(3)
userSubsetRecs = model.recommendForUserSubset(users, 10)
display(userSubsetRecs)

userId recommendations 471 List(List(177209, 5.2675586), List(107434, 4.963147), List(189473, 4.7630672), List(192261, 4.722142), List(157789, 4.7133274), List(190707, 4.6376963), List(146724, 4.602815), List(144202, 4.563069), List(166812, 4.542925), List(143422, 4.4984303)) 463 List(List(185659, 3.5780106), List(177209, 3.4935086), List(107434, 3.450104), List(155923, 3.4241645), List(187873, 3.4241645), List(173871, 3.3588603), List(177325, 3.3480835), List(155713, 3.3258576), List(135057, 3.3188667), List(171777, 3.3074894)) 148 List(List(177209, 5.6212225), List(107434, 5.4577627), List(157789, 5.1778603), List(155923, 5.1658907), List(187873, 5.1658907), List(185659, 5.1535153), List(177325, 5.116991), List(173871, 5.053779), List(144202, 5.0441837), List(192261, 5.0390935))

In [32]:
# Generate top 10 user recommendations for a specified set of movies
movieRecs = model.recommendForAllItems(10)
movies = ratings.select(als.getItemCol()).distinct().limit(3)
movieSubSetRecs = model.recommendForItemSubset(movies, 10)
display(movieSubSetRecs)

movieId recommendations 1580 List(List(53, 4.703871), List(543, 4.5154057), List(12, 4.34483), List(276, 4.302899), List(452, 4.268827), List(43, 4.2588677), List(93, 4.225232), List(413, 4.171688), List(523, 4.1124067), List(475, 4.1022787)) 3175 List(List(53, 4.6932282), List(43, 4.4675694), List(12, 4.3700953), List(276, 4.3195143), List(452, 4.248333), List(93, 4.1994324), List(1, 4.162988), List(99, 4.1585407), List(371, 4.1501093), List(171, 4.127697)) 2366 List(List(53, 4.3477855), List(236, 4.312768), List(122, 4.1948223), List(251, 4.1464543), List(371, 4.1429095), List(375, 4.1326103), List(171, 4.1321754), List(515, 4.121808), List(595, 4.1216), List(1, 4.1112175))

###Summary

Recommendation model was trained based on ALS aglorithm using 80% of the data, 20% of data was used to evaluate the model. 

The RMSE is 0.87, 17.4% out of highest rating of 5. To improve model performance, hyperparameters including # of ranks (latent factors), # of interations and regularization parameters can be tuned. Tuning these parameters can balance the bias error and variance error, also overcome the overfitting of the model. To do this, we can apply grid search and k-fold cross validation to select the best model. 

#####Explicit Data vs Implicit Data

In this project, explicit data is provided for analyzed. Explicit data is data where we have some sort of rating. Like the 1 to 5 ratings from the MovieLens or Netflix dataset. Here we know how much a user likes or dislikes an item, but this data is hard to come by. Your users might not spend the time to rate items or your app might not work well with a rating approach in the first place. It is common in many real-world use cases to only have access to implicit feedback (e.g. views, clicks, purchases, likes, shares etc.). These implicit feedbacks can be related to the preference of users on item. The model then tries to find latent factors that can be used to predict the expected preference of a user for an item.